Sources/XiEditor/Core/CoreNotification.swift

```diff
@@ -0,0 +1,243 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+extension Array where Element == [String: Any] {
+    /// A variant of the new built-in compactMap.
+    /// This variant returns an array containing the non-nil results of calling
+    /// the given transformation with each element of this sequence.
+    /// If any transform fails and returns nil, then the entire operation fails
+    /// and returns nil.
+    func xiCompactMap<ElementOfResult>(
+        _ transform: (Element) throws -> ElementOfResult?
+    ) rethrows -> [ElementOfResult]? {
+
+        // First transform all the items
+        let cmds = try! self.map(transform)
+            .filter { $0 != nil }
+
+        // Next check if any failed
+        // If we filter our transformed list and there's fewer than the original
+        if (cmds.count != self.count) {
+            return nil
+        }
+
+        // Now return the array as a non-optional array of transformed results
+        return cmds.map { $0! }
+    }
+}
+
+/// A notification from core.
+enum CoreNotification {
+    case alert(message: String)
+
+    case updateCommands(viewIdentifier: ViewIdentifier, plugin: String, commands: [Command])
+
+    case scrollTo(viewIdentifier: ViewIdentifier, line: Int, column: Int)
+
+    case addStatusItem(viewIdentifier: ViewIdentifier, source: String, key: String, value: String, alignment: String)
+    case updateStatusItem(viewIdentifier: ViewIdentifier, key: String, value: String)
+    case removeStatusItem(viewIdentifier: ViewIdentifier, key: String)
+
+    case update(viewIdentifier: ViewIdentifier, params: UpdateParams)
+
+    case configChanged(viewIdentifier: ViewIdentifier, config: Config)
+
+    case defStyle(params: DefStyleParams)
+
+    case availablePlugins(viewIdentifier: ViewIdentifier, plugins: [Plugin])
+    case pluginStarted(viewIdentifier: ViewIdentifier, pluginName: String)
+    case pluginStopped(viewIdentifier: ViewIdentifier, pluginName: String)
+
+    case availableThemes(themes: [String])
+    case themeChanged(name: String, theme: Theme)
+
+    case availableLanguages(languages: [String])
+    case languageChanged(viewIdentifier: ViewIdentifier, languageIdentifier: String)
+
+    case showHover(viewIdentifier: ViewIdentifier, requestIdentifier: Int, result: String)
+
+    case findStatus(viewIdentifier: ViewIdentifier, status: [FindStatus])
+    case replaceStatus(viewIdentifier: ViewIdentifier, status: ReplaceStatus)
+
+    static func fromJson(_ json: [String: Any]) -> CoreNotification? {
```

Comment:
This would be where all the parsing/verification of core notification JSON happens now. If this function returns a `CoreNotification` the rest of the app can depend on the notification being valid.

Reply:
Cool, I like this!